src/nmod_poly/mullow_fft_small.c

```diff
@@ -0,0 +1,517 @@
+   something much larger. This check is not needed for correctness;
+   it might furthermore want disabling if we care about often
+   multiplying structured with much smaller norms than generic ones. */
+static const int repack_limit_tab[] = {
+    0,
+    264497,  /* 1 */
+    256720,  /* 2 */
+    77049,  /* 3 */
+    36925,  /* 4 */
+    21385,  /* 5 */
+    14099,  /* 6 */
+    9872,  /* 7 */
+    7337,  /* 8 */
+    5565,  /* 9 */
+    4481,  /* 10 */
+    3609,  /* 11 */
+    3084,  /* 12 */
+    2562,  /* 13 */
+    2129,  /* 14 */
+    1859,  /* 15 */
+    1687,  /* 16 */
+    1446,  /* 17 */
+    1264,  /* 18 */
+    1160,  /* 19 */
+    1035,  /* 20 */
+    942,  /* 21 */
+    874,  /* 22 */
+    725,  /* 23 */
+    662,  /* 24 */
+    626,  /* 25 */
+    587,  /* 26 */
+    562,  /* 27 */
+    517,  /* 28 */
+    490,  /* 29 */
+    426,  /* 30 */
+    410,  /* 31 */
+    363,  /* 32 */
+    342,  /* 33 */
+    336,  /* 34 */
+    321,  /* 35 */
+};
+
+/* Alternative table with cutoffs tuned for more sparse polynomials. */
+#if 0
+    264497,  /* 1 */
+    264497,  /* 2 */
+    80976,  /* 3 */
+    39194,  /* 4 */
+    22473,  /* 5 */
+    16039,  /* 6 */
+    11338,  /* 7 */
+    7709,  /* 8 */
+    6647,  /* 9 */
+    4945,  /* 10 */
+    3904,  /* 11 */
+    3574,  /* 12 */
+    2770,  /* 13 */
+    2690,  /* 14 */
+    2213,  /* 15 */
+    1841,  /* 16 */
+    1607,  /* 17 */
+    1474,  /* 18 */
+    1339,  /* 19 */
+    1228,  /* 20 */
+    1805,  /* 21 */
+    1562,  /* 22 */
+    1517,  /* 23 */
+    1432,  /* 24 */
+    1252,  /* 25 */
+    1138,  /* 26 */
+    1085,  /* 27 */
+    978,  /* 28 */
+    933,  /* 29 */
+    858,  /* 30 */
+    826,  /* 31 */
+    711,  /* 32 */
+    698,  /* 33 */
+    680,  /* 34 */
+    620,  /* 35 */
+    587,  /* 36 */
+    552,  /* 37 */
+    494,  /* 38 */
+    474,  /* 39 */
+    478,  /* 40 */
+    406,  /* 41 */
+    414,  /* 42 */
+    406,  /* 43 */
+    396,  /* 44 */
+    360,  /* 45 */
+#endif
+
+/* Lemire, Kaser & Kurz method to compute x mod d for 32-bit d using
+   a single multiplication.
+
+   The #if-ed version is specialized for half-length (~16-bit moduli).
+   This is priori correct for all 16-bit x and d, but we actually need x all
+   the way up to 17 bits. Fortunately, it can be checked by brute force to be
+   correct for the range of d we need (e.g. 17-bit x and 15-bit d are fine).
+   It is easy to find counterexamples by choosing larger x or d,
+   e.g. x = 130119, d = 33096 fails.
+
+   We currently disable this version because it is actually benchmarks
+   worse on Zen 3 for unknown reasons; perhaps manual SIMD code is needed
+   for best performance.
+*/
+
+#if 1
+
+FLINT_FORCE_INLINE
+uint32_t u32_mod_preinv(uint32_t x, uint32_t d, uint64_t dinv)
+{
+    uint64_t l = dinv * x;
+
+#if (defined(_MSC_VER))
+    uint64_t hi, lo;
+    umul_ppmm(hi, lo, l, d);
+    return hi;
+#else
+    return ((__uint128_t) l * d ) >> 64;
+#endif
```

Comment:
```suggestion
#if !defined(__GNUC__)
    uint64_t hi, lo;
    umul_ppmm(hi, lo, l, d);
    return hi;
#else
    return ((__uint128_t) l * d ) >> 64;
#endif
```

Reply:
Indeed. BTW, we ought to add an ``n_mulhi`` function and reinvestigate basing this and ``umul_ppmm`` on intrinsics or ``uint128`` instead of inline assembly when the compiler supports this well; I get measurably better performance here on GCC with the ``uint128`` version.

Reply:
Hmm, yes.  And GCC is probably smart enough to do all sort of optimizations and such on `uint128_t`, so I can probably do `add_ssaaaa` with that as well in `longlong_asm_gcc.h`.